### ISSUE TITLE
Editor Completion: Consider not using null character to indicate Invoke [What are the actual ramifications?]

### ISSUE THREAD

**Issue by dpoeschl** [dotnet/roslyn#27426]
// TODO

**Version Used**: 

**Steps to Reproduce**:

1. 
2. 
3. 

**Expected Behavior**:

**Actual Behavior**:

**Comment by ivanbasov** [comment]
@AmadeusW 

In `TryGetApplicableToSpan`, we use `typeChar != 0` to exclude  `Invoke `or `InvokeAndCommitIfUnique`.
Can we have a better check for this condition? Maybe `TryGetApplicableToSpan` should accept a trigger type?

**Comment by AmadeusW** [comment]
Is there any reason to not use the character? My only reservation is about changing the API that is live in 15.8

**Comment by CyrusNajmabadi** [comment]
>  is there any reason to not use the character?

Yes.  It doesn't make a sense.  :)

If a user hits enter, let me know that enter was hit.  if they double-clicked the item in the list, let me know.   trying to map all inputs into the domain of some other type is somewhat abusive.

**Comment by AmadeusW** [comment]
I'm not disagreeing. We should have had this discussion over half a year ago when the API was in design and development. Back then, we all decided to make a concise API where the language service doesn't need to know all the details. Right now I must mark this as won't fix because this is the shipped API.

**Comment by CyrusNajmabadi** [comment]
> We should have had this discussion over half a year ago when the API was in design and development.

That would have been nice :)   But i'm curious why the API shipped before we knew if it would meet the needs of hte the LS powering C#/VB/JS/TS :-/

> Back then, we all decided to make a concise API where the language service doesn't need to know all the details. 

So, IMO, an API that combines these concepts is one where the LS "needs to know all the details".  Because the LS now needs to know things like "i have to handle the null character".  The way someone doesn't have to know the details is to make the API prescriptive.  So the API would actually say: "implement this for when the user hits enter", versus "implement this for when the user hits a letter."

**Comment by AmadeusW** [comment]
> But i'm curious why the API shipped before we knew if it would meet the needs of hte the LS powering C#/VB/JS/TS :-/

After a few months of meetings, emails and implementation, API was mature _enough_, yet LS were behind schedule. Roslyn won't use the API until dev16.

> The way someone doesn't have to know the details is to make the API prescriptive. So the API would actually say: "implement this for when the user hits enter", versus "implement this for when the user hits a letter."

That would have been good. Currently the API says "implement this for when user commits" and a character is an extra bit of information. Turns out it's not enough for a particular case of the LS.

The API says "implement this for when we are committing", and Editor is the one that says "we are comitting.

**Comment by CyrusNajmabadi** [comment]
> After a few months of meetings, emails and implementation, API was mature enough, yet LS were behind schedule. 

My personal feedback would be that in that case the API should either not be shipped, or should be shipped in beta ('subject to change') form.  The existing shipped languages have pushed these scenarios to very advanced places, and it would be really desirable to make sure hte replacements are suitable for these primary customers before shipping.

> Roslyn won't use the API until dev16.

My personal feedback would be that that means the API should have been in beta form until that point in time.  :)

> Turns out it's not enough for a particular case of the LS.

Right.  So this goes against the 'mature' claim.

Note: at this point, i suppose nothing can be done.  But i think can be used as a good learning example for future cross-team work.  I think it can be fine to ship early (especially to garner third party feedback).  However, it's really difficult to ship early, then discover later that there are issues, but then feel locked in because things were already shipped.

**Comment by AmadeusW** [comment]
The situation not as dire as you think, where nothing can be done. We have discussed these scenarios over and over again, and here is a recap:

The purpose of `TryGetApplicableToSpan` is to produce a span that would be pertinent to completion at a given location. This method is supposed to return as quickly as possible. 

`GetCompletionContextAsync` is an appropriate place to return items of your choice, based on user action. User action is provided as a parameter. This is the method where you want to handle _very advanced places_.

We discussed this numerous times with @dpoeschl and it sounds like Roslyn is a refactoring away from not needing the type char in TryGetApplicableToSpan.
